src/libstore/build/goal.hh

```diff
@@ -98,13 +99,99 @@ struct Goal : public std::enable_shared_from_this<Goal>
      */
     ExitCode exitCode = ecBusy;
 
-protected:
+public:
     /**
      * Build result.
      */
     BuildResult buildResult;
 
-public:
+    /*
+     * Suspend our goal and wait until we get work()-ed again.
+     */
+    struct SuspendGoal {};
+    struct Return {};
+    struct promise_type;
+    using handle_type = std::coroutine_handle<promise_type>;
+    // FIXME: Allocate explicitly on stack since HALO thing doesn't really work,
+    // specifically, there's no way to uphold the requirements when trying to do
+    // tail-calls without using a trampoline AFAICT.
+    // NOTES:
+    // These are good resources for understanding how coroutines work:
+    // https://lewissbaker.github.io/
+    // https://www.chiark.greenend.org.uk/~sgtatham/quasiblog/coroutines-c++20/
+    // https://www.scs.stanford.edu/~dm/blog/c++-coroutines.html
+    struct [[nodiscard]] Co {
+        handle_type handle;
+        explicit Co(handle_type handle) : handle(handle) {};
+        Co(const Co&) = delete;
+        Co &operator=(const Co&) = delete;
+        void operator=(Co&&);
+        Co(Co&& rhs);
+        ~Co();
+
+        bool await_ready() { return false; };
+        std::coroutine_handle<> await_suspend(handle_type handle);
+        void await_resume() {};
```

Comment:
The ones that don't have descriptions in the file (which should be moved here) should gain new doxygens